spec/app/domain/authentication/authenticate_spec.rb

```diff
@@ -0,0 +1,172 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+RSpec.describe 'Authentication::Authenticate' do
+  ####################################
+  # Available Authenticators - doubles
+  ####################################
+
+  def authenticator(pass:)
+    double('Authenticator').tap do |x|
+      allow(x).to receive(:valid?).and_return(pass)
+    end
+  end
+
+  def input(
+    authenticator_name: 'authn-always-pass',
+    service_id: nil,
+    account: 'my-acct',
+    username: 'my-user',
+    password: 'my-pw',
+    origin: '127.0.0.1',
+    request: nil
+  )
+    Authentication::Input.new(
+      authenticator_name: authenticator_name,
+      service_id: service_id,
+      account: account,
+      username: username,
+      password: password,
+      origin: origin,
+      request: request
+    )
+  end
+
+  let (:authenticators) do
+    {
+      'authn-always-pass' => authenticator(pass: true),
+      'authn-always-fail' => authenticator(pass: false)
+    }
+  end
+
+  ####################################
+  # Security doubles
+  ####################################
+
+  let (:mocked_security_validator) {double("MockSecurityValidator")}
+  let (:mocked_origin_validator) {double("MockOriginValidator")}
+
+  before(:each) do
+    allow(Authentication::ValidateSecurity)
+      .to receive(:new)
```

Comment:
Ok, this is the first real conceptual error I've found so far.  We never want to do our mocks like this, where we are changing the behavior of a class name to act differently from how it would in production.  

In fact, in many ways the whole point of the command class dependencies is so that we can inject mock objects directly into the command class constructors in tests.  That is, the interface exists so we don't have to do this "Make the hardcoded thing act different in tests" hack.   Stuff like `Authentication::ValidateSecurity.new` in the production just gives our command classes _default_ dependencies so our production client code doesn't have to explicitly name them everywhere.  But we do expect to have to explicitly override these defaults in the tests.

Reply:
cool, I didn't know it's possible to override a dependency in your CommandClass. Refactored this throughout the code